core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java

```diff
@@ -272,29 +273,21 @@ private BitSet getFilteredIndexesForMeasures(ColumnPage columnPage,
     // Get the measure values from the chunk. compare sequentially with the
     // the filter values. The one that matches sets it Bitset.
     BitSet bitSet = new BitSet(rowsInPage);
-    Object[] filterValues = msrColumnExecutorInfo.getFilterKeys();
-
-    SerializableComparator comparator = Comparator.getComparatorByDataTypeForMeasure(msrType);
+    Set filterValuesSet = msrColumnExecutorInfo.getFilterKeysSet();
     BitSet nullBitSet = columnPage.getNullBits();
-    for (int i = 0; i < filterValues.length; i++) {
-      if (filterValues[i] == null) {
-        for (int j = nullBitSet.nextSetBit(0); j >= 0; j = nullBitSet.nextSetBit(j + 1)) {
-          bitSet.set(j);
-        }
-        continue;
-      }
-      for (int startIndex = 0; startIndex < rowsInPage; startIndex++) {
-        if (!nullBitSet.get(startIndex)) {
-          // Check if filterValue[i] matches with measure Values.
-          Object msrValue = DataTypeUtil
-              .getMeasureObjectBasedOnDataType(columnPage, startIndex,
-                  msrType, msrColumnEvaluatorInfo.getMeasure());
-
-          if (comparator.compare(msrValue, filterValues[i]) == 0) {
-            // This is a match.
-            bitSet.set(startIndex);
-          }
+    for (int startIndex = 0; startIndex < rowsInPage; startIndex++) {
+      if (!nullBitSet.get(startIndex)) {
+        // Check if filterValue[i] matches with measure Values.
+        Object msrValue = DataTypeUtil
+            .getMeasureObjectBasedOnDataType(columnPage, startIndex,
+                msrType, msrColumnEvaluatorInfo.getMeasure());
+
+        if (filterValuesSet.contains(msrValue)) {
```

Comment:
HashSet/Map lookup depends on hashcode . I think this may not match incase of decimal & float value  Example. 123.000 & 123.0  comparison with equals would work but hashcode would be different.

I think the optimal approach would be to use TreeMap with  comparator by data type [TreeMap(Comparator<? super K> comparator)] which does binary search or sorted filter values.

Reply:
I debuged case of decimal and float, if one filed's type is  float and one of values is  123.000 and filter condition is 123.00,  in picture's place the filed's value was converted to 123.0 and filter's value was converted to 123.0, so they are equal, the decimal is the same.

Reply:
Ok.. Float and Double hashcode implementation is taken care. But Decimal data type seems creates problem. Please refer below sample code.

BigDecimal bigDecimal = new BigDecimal("4743.00");
    BigDecimal  bigDecimal2 = new BigDecimal("4743.0");

    System.out.println(bigDecimal.compareTo(bigDecimal2)==0); // compareTo returns true

    HashSet<BigDecimal> decimals = new HashSet<>();
    decimals.add(bigDecimal);
    decimals.add(bigDecimal2);

    System.out.println(decimals.toString()); // Output: [4743.00, 4743.0]
    System.out.println(decimals.contains(new BigDecimal("4743"))); //Returns false
    System.out.println(decimals.contains(new BigDecimal("4743.0000"))); // Returns false
    System.out.println(decimals.contains(new BigDecimal("4743.00"))); // Returns true
    System.out.println(decimals.contains(new BigDecimal("4743.0"))); // Returns true

    TreeSet<BigDecimal> treeSet = new TreeSet();
    treeSet.add(bigDecimal);
    treeSet.add(bigDecimal2);

    System.out.println(treeSet.toString()); // Output:  [4743.00]
    System.out.println(treeSet.contains(new BigDecimal("4743"))); //Returns true
    System.out.println(treeSet.contains(new BigDecimal("4743.0000"))); // Returns true
    System.out.println(treeSet.contains(new BigDecimal("4743.00"))); // Returns true
    System.out.println(treeSet.contains(new BigDecimal("4743.0"))); // Returns true

So, please add a test case to validate this scenario also.

Reply:
I agree with @KanakaKumar current implementation is based on `SerializableComparator ` so it is better to use `TreeSet `  to use `compareTo` instead of 'hashcode' and 'equals' methods. `BigDecimal` behaviour is different for `equal` and `compareTo` methods. So better stick with `TreeSet` to continue the old behaviour rather than changing the behaviour.

Reply:
if you guys review this code without context, it exactly is  what you said above,  but the filter value and colume value already was converted to be that type is the same and precision is the same before my modified code. so the situation you worry about will never happen. I add a test case to the project, and it all passed. the last point : if I use treemap, it will increase the time complexity ,please review again, thank you,

Reply:
how about this PR,does need to do modify again?

Reply:
@zhxiaoping , OK. Seems right now IncludeFilterExecuterImpl is not used for Decimal type case.  
Please correct the test case with rows having different  precision (4743, 4743.00, 4743.0000, 4743.0)  and assert the IN filter query should result all values to identify failures in case its used in future for BigDecimals.